### PR TITLE
Add support for Redis Sentinel by using ioredis npm package. Closes #1769

### LINKED ISSUE #1769: Add support for redis sentinel
We are using Redis Sentinal at our company and therefore cube.js cannot use the Redis server(s)

I propose we use ioredis like described in https://github.com/cube-js/cube.js/issues/146#issuecomment-591085482 by @keeganmccallum

We use ioredis for other services at our company and it seems to work well

```
const Redis = require('ioredis');

module.exports = function createRedisClient(url) {
    // map exec responses to match node-redis: https://github.com/luin/ioredis/wiki/Migrating-from-node_redis
    Redis.Pipeline.prototype.execAsync = function () {
        return new Promise((resolve, reject) => {
            // rely on scope
            this.exec(function (err, res) {
                console.log(err, res)
                if (err) return reject(err);
                return resolve(res.map((nestArr) => nestArr[1]));
            })
        })
    };
    const [host, port] = url.replace('redis://', '').split(':');
    let options = {
        sentinels: [{ host, port }],
        name: "mymaster"
    };

    if (process.env.REDIS_TLS === 'true') {
        options.tls = {};
    }

    if (process.env.REDIS_PASSWORD) {
        options.password = process.env.REDIS_PASSWORD;
    }

    const client = new Redis(options);

    ['brpop', 'del', 'get', 'hget', 'rpop', 'set', 'zadd', 'zrange', 'zrangebyscore', 'keys'].forEach(
        k => {
            client[`${k}Async`] = client[k];
        }
    );

    return client;
};
```

### PR DESCRIPTION
**Pull request by jrylander**
[Add support for redis sentinel #1769]

[Use ioredis]

Changes from npm package redis to ioredis. Since ioredis supports promises directly, code related to creating and calling async methods is removed.

### CODE REVIEW

**Comment by jrylander** [comment]
I have not added the setup of a sentinel redis but the redis integration tests run as usual.

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-10.17
+12.20.1
```
Comment (ovr) [inline .nvmrc:1]: @jrylander Why did you change the default version for local development? Do you have any limitations with 10? We still supported Node.js 8+...

File: packages/cubejs-query-orchestrator/package.json [hunk 2]
```diff
@@ -33,14 +33,14 @@
...
-    "redis": "^3.0.2"
+    "ioredis": "^4.19.4",
+    "ramda": "^0.27.0"
   },
   "devDependencies": {
     "@cubejs-backend/linter": "^0.25.0",
     "@types/generic-pool": "^3.1.9",
+    "@types/ioredis": "^4.17.11",
     "@types/jest": "^26.0.15",
-    "@types/node": "^8.10.66",
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/package.json:43]: What is a reason to removing @types/node for this package?

File: .nvmrc (hunk 1 above)
Comment (jrylander) [reply .nvmrc:1]: I looked at what is used in dev.Dockerfile and thought synching those would be a good idea. I see your point, should dev.Dockerfile still use 12?

File: packages/cubejs-query-orchestrator/package.json (hunk 2 above)
Comment (jrylander) [reply packages/cubejs-query-orchestrator/package.json:43]: My bad. I fiddled with versions of node. Thank you for noticing.

File: .nvmrc (hunk 1 above)
Comment (ovr) [reply .nvmrc]: Oh, ok. Forget about that. let's use 12 then.

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisFactory.ts [hunk 3]
```diff
@@ -64,10 +22,5 @@ export function createRedisClient(url: string, opts: ClientOpts = {}) {
     options.password = process.env.REDIS_PASSWORD;
   }
 
-  return decorateRedisClient(
-    redis.createClient({
-      ...options,
-      ...opts,
-    })
-  );
+  return Promise.resolve(new Redis(options));
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/orchestrator/RedisFactory.ts:25]: I think it's not needed to wrap `new Redis` with `Promise.resolve` because `createRedisClient` function already declared as `async` and it will use `AsyncFunction` type which always return Promise on call.

Maybe I miss something.

Comment (jrylander) [reply packages/cubejs-query-orchestrator/src/orchestrator/RedisFactory.ts]: No, it is not needed. The async will add it

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts [hunk 4]
```diff
@@ -1,23 +1,22 @@
+import Redis from 'ioredis';
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts:1]: I think, It's better to import redis directly, instead of using default export.

```suggestion
import { Redis } from 'ioredis';
```

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts [hunk 5]
```diff
@@ -1,23 +1,22 @@
...
-
-export type CreateRedisClientFn = () => PromiseLike<AsyncRedisClient>;
-
+export type CreateRedisClientFn = () => PromiseLike<Redis.Redis>;
 export interface RedisPoolOptions {
   poolMin?: number;
   poolMax?: number;
   createClient?: CreateRedisClientFn;
-  destroyClient?: (client: AsyncRedisClient) => PromiseLike<void>;
+  destroyClient?: (client: Redis.Redis) => PromiseLike<void>;
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts:10]: To use it like `Redis`, instead of `Redis.Redis`

```suggestion
  destroyClient?: (client: Redis) => PromiseLike<void>;
```

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts [hunk 6]
```diff
@@ -1,23 +1,22 @@
...
 }
 
 const MAX_ALLOWED_POOL_ERRORS = 100;
 
 export class RedisPool {
-  protected readonly pool: Pool<AsyncRedisClient>|null = null;
+  protected readonly pool: Pool<Redis.Redis>|null = null;
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts:16]: To use it like `Redis`, instead of `Redis.Redis`

```suggestion
  protected readonly pool: Pool<Redis>|null = null;
```

**Review by ovr — APPROVED** [review]
Added few minor comments, LGTM

**Comment by jrylander** [comment]
I am getting errors after a while on our redis sentinel.

**Comment by jrylander** [comment]
I will put this version into internal use at SVT and see if it is stable enough

**Comment by paveltiunov** [comment]
@jrylander Hey Johan! Thanks for contributing it! Seems to me like a major change. I'd suggest enabling it by env flag for now and keep the existing driver implementation as a default one. This way we can merge it without fear of breaking existing deployments. cc @ovr

**Comment by jrylander** [comment]
> @jrylander Hey Johan! Thanks for contributing it! Seems to me like a major change. I'd suggest enabling it by env flag for now and keep the existing driver implementation as a default one. This way we can merge it without fear of breaking existing deployments. cc @ovr

It is probably a good idea in order to get it released. I do really hope we can scrap the current implementation so there is not two to keep updated. I'll start the work to adapt the ioredis to behave like current one does (*Async et al).

Thank you for getting back with feedback, everyone!

**Comment by jrylander** [comment]
Sorry for messing with this! I am still getting errors after some time of having our redis sentinel connection up. Doing it in spare time so bare with me.

**Comment by jrylander** [comment]
It seems to work with ioredis and pooling now

**Review by paveltiunov — COMMENTED** [review]
@jrylander Hey Johan! You did a really great job here! There're a couple of things to polish and we're good to go! Thanks for contributing this!

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisSentinelFactory.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
+import Redis, { Redis as redis, RedisOptions, Pipeline } from 'ioredis';
+
+import config from '../config';
+
+function debugLog(msg) {
```
[the remark this answers is not in the document]
Comment (jrylander) [reply packages/cubejs-query-orchestrator/src/orchestrator/RedisSentinelFactory.ts:5]: I tried to find how to log things, can you give me a pointer?

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts [hunk 8]
```diff
@@ -1,12 +1,26 @@
...
 import genericPool, { Pool, Options as PoolOptions } from 'generic-pool';
+import AsyncRedisClient from './AsyncRedisClient';
+import RedisClientOptions from './RedisClientOptions';
+import { createRedisClient as createNodeRedisClient } from './RedisFactory';
+import { createRedisSentinelClient } from './RedisSentinelFactory';
+import config from '../config';
 
-import { createRedisClient, AsyncRedisClient } from './RedisFactory';
+function createRedisClient(url: string, opts: RedisClientOptions = {}) {
+  if (config.FLAG_ENABLE_REDIS_SENTINEL) {
```
[the remark this answers is not in the document]
Comment (jrylander) [reply packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts:10]: I changed to CUBEJS_REDIS_USE_IOREDIS and for the sentinel address CUBEJS_REDIS_SENTINEL

[the remark this answers is not in the document]
Comment (jrylander) [reply packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts:10]: Hopefully ioredis works and we can retire redis npm package so we do not have two packages for the same thing

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisSentinelFactory.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  client.on('close', () => {
+    debugLog('Redis connection closed');
+  });
+
+  client.on('end', () => {
+    debugLog('Redis connection ended');
+  });
+
+  client.on('error', (e) => {
+    console.error('Redis connection failed: ', e);
```
[the remark this answers is not in the document]
Comment (jrylander) [reply packages/cubejs-query-orchestrator/src/orchestrator/RedisSentinelFactory.ts:57]: Errors are thrown by the client but I removed the error log which I only used for debugging anyways

**Comment by jrylander** [comment]
> @jrylander Hey Johan! You did a really great job here! There're a couple of things to polish and we're good to go! Thanks for contributing this!

Removed the logs. Only neccessary when debugging really.

File: packages/cubejs-query-orchestrator/src/config.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+function envIntWithDefault(envVariable: string, defaultValue: number) {
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/config.ts:1]: It's better to use `getEnv` from `@cubejs-backend/shared` instead of doing one more place where we are getting and verifying env variables.

Comment (jrylander) [reply packages/cubejs-query-orchestrator/src/config.ts]: Done. (I did not find it before you pointed me to it - tnx)

**Comment by igorlukanin** [comment]
@hassankhan Would you please review the changes to the docs? (Great job, @jrylander!)

**Comment by jrylander** [comment]
> @hassankhan Would you please review the changes to the docs? (Great job, @jrylander!)

Thanks for pointing it out. It seems you do a very good job in keeping the quality up on this software and I thank you for it.

File: docs/content/Deployment/Production-Checklist.md [hunk 11]
```diff
@@ -39,6 +39,14 @@ password, please set it via the `REDIS_PASSWORD` environment variable. Set the
 connections. Ensure your Redis cluster allows at least 15 concurrent
 connections.
 
+### Redis Sentinel
+
+Redis Sentinel support via ioredis was recently added. In order to activate
```
Comment (hassankhan) [inline docs/content/Deployment/Production-Checklist.md:44]: Could we add a link to the ioredis repo here like the above suggestion? Just so people who are unaware of it can quickly brush up on it 😄

File: docs/content/Deployment/Production-Checklist.md [hunk 12]
```diff
@@ -39,6 +39,14 @@ password, please set it via the `REDIS_PASSWORD` environment variable. Set the
...
+this feature, set the `CUBEJS_REDIS_USE_IOREDIS` environment variable to
+`true` and then set `CUBEJS_REDIS_SENTINEL` to to the address and port
```
Comment (hassankhan) [inline docs/content/Deployment/Production-Checklist.md:46]:
```suggestion
`true` and then set `CUBEJS_REDIS_SENTINEL` to the address and port
```

File: docs/content/Deployment/Production-Checklist.md [hunk 13]
```diff
@@ -39,6 +39,14 @@ password, please set it via the `REDIS_PASSWORD` environment variable. Set the
...
+(`host:port`) to allow Cube.js to connect to Redis Sentinel. `REDIS_URL` environment
```
Comment (hassankhan) [inline docs/content/Deployment/Production-Checklist.md:47]:
```suggestion
(`host:port`) to allow Cube.js to connect to Redis Sentinel. The `REDIS_URL` environment
```

File: docs/content/Configuration/Environment-Variables-Reference.md [hunk 14]
```diff
@@ -34,6 +34,8 @@ options][link-config].
...
 | `CUBEJS_REDIS_POOL_MAX`              | The maximum number of connections to keep active in the Redis connection pool. Must be higher than `CUBEJS_REDIS_POOL_MIN`. Defaults to `1000`                                                     | A valid number of connections.                            |
 | `CUBEJS_REDIS_POOL_MIN`              | The minimum number of connections to keep active in the Redis connection pool. Must be lower than `CUBEJS_REDIS_POOL_MAX`. Defaults to `2`                                                         | A valid number of connections                             |
+| `CUBEJS_REDIS_USE_IOREDIS`           | Makes the redis connection use ioredis instead of redis npm package. This flag must be true in order for connections to Redis Sentinel to work as well as connections directly to Redis. Defaults to `false`                                                         | `true`, `false`                             |
```
Comment (hassankhan) [inline docs/content/Configuration/Environment-Variables-Reference.md:37]:
```suggestion
| `CUBEJS_REDIS_USE_IOREDIS`           | Use [`ioredis`][gh-ioredis] instead of[ `redis`][gh-node-redis]. This flag must be true in order for connections to Redis Sentinel to work as well as connections directly to Redis. Defaults to `false` | `true`, `false`                             		|
```

File: docs/content/Configuration/Environment-Variables-Reference.md [hunk 15]
```diff
@@ -34,6 +34,8 @@ options][link-config].
...
+| `CUBEJS_REDIS_SENTINEL`              | The host URL for a Redis Sentinel server                                                         | A valid Redis host. May include port                             |
 
 [link-tz-database]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
```
Comment (hassankhan) [inline docs/content/Configuration/Environment-Variables-Reference.md:40]:
```suggestion
[gh-ioredis]: https://github.com/luin/ioredis
[gh-node-redis]: https://github.com/NodeRedis/node-redis
[link-tz-database]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
```

File: docs/content/Configuration/Environment-Variables-Reference.md (hunk 14 above)
Comment (hassankhan) [reply docs/content/Configuration/Environment-Variables-Reference.md:37]: I've added inline links to the `ioredis` and `redis` repositories, the URL referenced is in the next suggestion

**Review by hassankhan — CHANGES_REQUESTED** [review]
Awesome work @jrylander 🙌 

I've left a few comments, please do let me know if they don't make sense 😄

File: docs/content/Deployment/Production-Checklist.md (hunk 13 above)
Comment (jrylander) [reply docs/content/Deployment/Production-Checklist.md:47]: Nice

File: docs/content/Deployment/Production-Checklist.md (hunk 11 above)
Comment (igorlukanin) [reply docs/content/Deployment/Production-Checklist.md:44]: I would also suggest a few ideas:
* Docs are usually considered a long-lasting and time-insensitive medium (quite the opposite of a blog post or a changelog entry) meaning that it's often unclear what "recently" in docs really means 😄. Maybe we can skip it here?
* Providing links to Redis Sentinel and ioredis is a great idea but I'd also suggest to add a few words about the value of using them. People who are already familiar with this software don't need this but people who'd like to have a HA Redis but just don't know it's called Sentinel might benefit from such clarification 😄

Comment (jrylander) [reply docs/content/Deployment/Production-Checklist.md:44]: Yup

File: docs/content/Deployment/Production-Checklist.md [hunk 16]
```diff
@@ -39,6 +39,16 @@ password, please set it via the `REDIS_PASSWORD` environment variable. Set the
...
+### Redis Sentinel
+
+Redis Sentinel support via the [`ioredis`][gh-ioredis] npm package was recently
+added. In order to activate this feature, set the `CUBEJS_REDIS_USE_IOREDIS`
+environment variable to `true` and then set `CUBEJS_REDIS_SENTINEL` to the
+address and port (`host:port`) to allow Cube.js to connect to the Redis
+Sentinel. `REDIS_URL` environment variable should not be set in this case.
+Ioredis is usable with redis as well and the choice of type is done by setting
+either `CUBEJS_REDIS_SENTINEL ` or `REDIS_URL`;
+
```
[the remark this answers is not in the document]
Comment (hassankhan) [reply docs/content/Deployment/Production-Checklist.md:51]: Just need to add the link otherwise it won't link anywhere 😀

**Review by hassankhan — CHANGES_REQUESTED** [review]
Minor followup fixes, looks great @jrylander 👍

**Review by hassankhan — APPROVED** [review]

**Comment by ovr** [comment]
Thank you @jrylander 🍰 🥮 

Landed in `master` in https://github.com/cube-js/cube.js/commit/a5e7972485fa97faaf9965b9794b0cf48256f484

Please use `REDIS_URL=redis+sentinel://localhost:26379,otherhost:26479/mymaster/5` to connect with sentinel.

Thanks
